fix(api-reference): add missing info, types and `<Since />`

ArmandPhilippot opened the pull request:
This PR is a follow-up to my discussion with @sarah11918 in #8929.

#### Description (required)

I updated `api-reference.mdx` to add missing (or to update) information like:
* `<Since />` component,
* defaults,
* types for individual properties.

For this task, I tried to keep track of my references in each commit (with URLs) so that you can verify the accuracy of the information (this is why this PR has so many commits, I found it was easier to divide the file in parts).

I left out some types that seemed too complex to understand by reading the documentation. Also, I modified a type but I'm not sure its definition is clear from reading the documentation. I'll add a review comment.

If I haven't added a `<Since />` component it's either because I haven't found when the property/component was added or because it dates from v1 or earlier.

Additionally, I noticed some missing sections:
* all the properties of `Astro.cookies` was described except `merge`, I added it
* `getDataEntryById` is missing but `getEntryBySlug` is marked as deprecated in the documentation (I couldn't find a reference about this) so I don't know if we should add it (I guess it is also deprecated)

Finally, I removed:
* `Astro.canonicalURL` because it seems it was removed in `v2.0.0` (see https://github.com/withastro/astro/pull/5707 and we are currently in `v4.12.1`)

#### Related issues & labels (optional)

- Related: #8929 + #6467
- Suggested label: improve-documentation

#### Documentation "coding" style related

While translating the doc in French and while updating this file, I noticed that some parts of the files was inconsistent, more exactly when we have `<Since />` and/or `**Types:**`:
* Sometimes `<Since />` OR `**Types:**` are wrapped in `<p>...</p>` tags, other times it is not the case
* When `<Since />` AND `**Types:**` are both provided, sometimes they are in a block wrapped with `<p>...</p>`, other times they are separated (often we have a `<Since />` wrapped in `<p>...</p>` tags followed by or put after `**Types:**` and separated by a new line)

While updating this file I wasn't sure what to do. In [Astro Docs Docs](https://contribute.docs.astro.build/reference/custom-components/#examples-1), we have an example but I wasn't sure if it was a guideline or just an example.

So if you have any guidance about the formatting, I'll be happy to update this PR to make the documentation more consistent (at least in this file)! 😄

astrobot-houston commented:
<!-- lunaria-action-comment -->

## Lunaria Status Overview


🌕 **This pull request will trigger status changes.**

<details>
<summary>Learn more</summary>

By default, every PR changing files present in the [Lunaria configuration's `files` property](https://lunaria.dev/reference/configuration/#files-required) will be considered and trigger status changes accordingly.

You can change this by adding one of the keywords present in the [`ignoreKeywords` property in your Lunaria configuration file](https://lunaria.dev/reference/configuration/#ignorekeywords) in the PR's title (ignoring all files) or by [including a tracker directive](https://lunaria.dev/guides/tracking/#tracker-directives) in the merged commit's description. 
</details>

### Tracked Files




| Locale | File                                                                                                                                                                     | Note                                                       |
| ------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ---------------------------------------------------------- |
| en     | [reference/api-reference.mdx](https://github.com/withastro/docs/blob/d46cbb891b568e979a5fb45b1a9f3f3e000c7574/src%2Fcontent%2Fdocs%2Fen%2Freference%2Fapi-reference.mdx) | Source changed, localizations will be marked as outdated.  |

<details>
<summary>Warnings reference</summary>

| Icon | Description                                                                                                                                       |
| ---- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| 🔄️  | The source for this localization has been updated since the creation of this pull request, make sure all changes in the source have been applied. |

</details>

ArmandPhilippot commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -1120,14 +1189,16 @@ This means collections **cannot** store a mix of content and data formats. You m
 #### `schema`
 
-**Type:** `TSchema extends ZodType`
+**Type:** `TSchema extends ZodType | (context: SchemaContext) => TSchema extends ZodType`
```
I wasn't sure how to update this type. Is `TSchema extends ZodType` really meaningful for someone new to Astro (or Typescript)? Reading this, it seems weird to me since the other types are usable while here, it is an indication for those who understand Typescript. Even if it's not a real type, shouldn't we just stick to `ZodType`?

```
`ZodType | (context: SchemaContext) => ZodType`
```

delucis replied:
Yes, I think your suggestion makes sense!

By the way, not required, but something we’ve done in the Starlight docs is actually link type definitions when we have them, i.e. here it might be:

```suggestion
**Type:** <code>ZodType | (context: <a href="#schemacontext">SchemaContext</a>) => ZodType</code>
```

ArmandPhilippot replied:
Nice idea, this can be useful to better understand the return type! I'll check the other types (I'm thinking in particular of `AstroCookie` ).

ArmandPhilippot commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -206,28 +214,30 @@ Astro.response.headers.set('Set-Cookie', 'a=b; Path=/;');
 
 <p>
-**Type:** `(key: string, options?: CookieGetOptions) => AstroCookie`
+**Type:** `(key: string, options?: CookieGetOptions) => AstroCookie | undefined`
```
The real type is `AstroCookieGetOptions`. Was the type renamed at some point (I couldn't find a reference) or was it to shorten the type declared in the documentation? (I found other examples where the `Astro` prefix is ​​missing)

delucis replied:
Again, I would say let’s use the full name. Very probable something was renamed and these docs were never updated.

ArmandPhilippot commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -52,14 +52,19 @@ Markdown files have the following interface:
   /* The rendered path of this file */
 	url: string | undefined;
   /* Astro Component that renders the contents of this file */
 	Content: AstroComponent;
```
I hesitated to update this type (I believe it is present elsewhere as well). `AstroComponent` does not exist, the real type is `AstroComponentFactory`. Should I update all the occurrences or has the type been simplified for documentation purposes?

delucis replied:
I think using the accurate name is not a bad idea. Admittedly `AstroComponentFactory` is not a type we really document, so either way it doesn’t matter **_too_** much, but might as well use the same name we use in the code base in case anyone does want to search for it.

sarah11918 commented:
Thank you @ArmandPhilippot , this looks great! I'm going to ask @delucis to review this one!

delucis commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -52,14 +52,19 @@ Markdown files have the following interface:
 	url: string | undefined;
   /* Astro Component that renders the contents of this file */
 	Content: AstroComponent;
+  /** Raw Markdown file content, excluding layout HTML and YAML frontmatter */
+	rawContent(): string;
+  /** Markdown file compiled to HTML, excluding layout HTML */
+	compiledContent(): string;
```
I believe these are only available for Markdown but **not** MDX. Do you think that should be mentioned here?

Similar to in the Markdown docs: https://docs.astro.build/en/guides/markdown-content/#exported-properties

ArmandPhilippot replied:
Indeed, they are only available for Markdown. I had hesitated to add them but since this is `MarkdownInstance` and not `MDXInstance` I figured they should appear. But a `Markdown only` comment makes perfect sense!

delucis commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -52,14 +52,19 @@ Markdown files have the following interface:
 export interface MarkdownInstance<T extends Record<string, any>> {
```
I’m wondering if we should mention `MarkdownInstance` in the introductory line above. Currently there are no search results for it because code samples are excluded from indexing.

For example something like

> Markdown files loaded with `Astro.glob()` return the following `MarkdownInstance` interface:

delucis commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -242,6 +252,14 @@ Invalidates a cookie by setting the expiration date in the past (0 in Unix time)
+**Type:** `(cookies: AstroCookies) => void`
+</p>
+
+Merges a new `AstroCookies` instance into the current instance. Any new cookies will be added to the current instance, overwriting any existing cookies with the same name.
```
I think this probably needs an “and” to separate the two clauses as it is not the _new_ cookies that overwrite, but only the ones matching existing cookies.

```suggestion
Merges a new `AstroCookies` instance into the current instance. Any new cookies will be added to the current instance and any cookies with the same name will overwrite existing values.
```

delucis commented on src/content/docs/en/reference/api-reference.mdx:
````diff
@@ -397,18 +415,12 @@ if (!isLoggedIn(cookie)) {
 ---
 ```
 
-### `Astro.canonicalURL`
-
-:::caution[Deprecated]
-Use [`Astro.url`](#astrourl) to construct your own canonical URL. 
-:::
-
-The [canonical URL][canonical] of the current page.
-
````
A note on this: we probably kept this around because there were some links to the `#astrocanonicalurl` heading slug. That said, I don’t think it’s still needed, because as you noted it was removed more than a year ago.

If we wanted to be responsible, we _could_ add a redirect for this case to `public/_redirects`:
```
/:lang:/reference/api-reference/#astrocanonicalurl   /:lang:/reference/api-reference/#astrourl
```

ArmandPhilippot replied:
The documentation doesn't seems to link to this section anymore, but external resources might so the redirection makes sense! I'll add it.

delucis reviewed (changes requested):
Wow, thank you @ArmandPhilippot for this amazingly thorough work! Love to see stuff like the default pagination values specified — really helpful 💜


> So if you have any guidance about the formatting, I'll be happy to update this PR to make the documentation more consistent (at least in this file)! 😄

The examples in Astro Docs Docs are correct! Ideally, when there are multiple short items like Type + `<Since>`, these should be grouped together rather than in separate paragraphs.

You can see this style in use in our [configuration reference](https://docs.astro.build/en/reference/configuration-reference/#redirects) which is generated automatically from code:

<img width="765" alt="Reference header reading “redirects” followed by type, default, and added in information grouped together visually" src="https://github.com/user-attachments/assets/82e085a3-9d5b-496b-b0a9-150c942d27a6">


> `getDataEntryById` is missing but `getEntryBySlug` is marked as deprecated in the documentation (I couldn't find a reference about this) so I don't know if we should add it (I guess it is also deprecated)

Yes, if I’m not mistaken, they are both deprecated in favour of `getEntry()`.

ArmandPhilippot commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -1279,13 +1449,14 @@ const draftBlogPosts = await getCollection('blog', ({ data }) => {
 
-<p><Since v="2.5.0" /></p>
+<p>
 
 **Types:**
-
 - `(collection: string, contentSlugOrDataId: string) => CollectionEntry<collection>`
 - `({ collection: string, id: string }) => CollectionEntry<collection>`
 - `({ collection: string, slug: string }) => CollectionEntry<collection>`
+<Since v="2.5.0" />
+</p>
```
I'm not sure about this one. Since we have a list of types the rendering might seem odd. I tried to place `<Since />` before `Types:` but we have a new line between the two. I think the Markdown parser ignores the `<p>` tag around because of the list inside. So I'm not sure what is the best option here.

delucis replied:
Oh boy, yes, had no idea we’d allowed someone to have function overloads like this 😅 

I _think_ it might look ok if `<Since>` appears first in this case?

<img width="753" alt="getEntry docs with added in version displayed above the list of valid types" src="https://github.com/user-attachments/assets/dfc0ef73-181c-4148-b9d5-b6e7c57ee5c3">

ArmandPhilippot replied:
I felt like the spacing was bigger when I tried... I must have tested too quickly. So yes, it seems to be the best option when using multiple types!

ArmandPhilippot replied:
It seems I'm struggling with this one. In all my tests, `Types:` is always rendered in its own `<p>` tag. So the spacing does not match your screenshot.

For example, the following should render both in a single `<p>` tag, but no... `Types:` is stubborn 😬 and has is own `<p>` tag in the end.

```mdx
<p>

<Since v="2.5.0" />
**Types:**
</p>
```

@delucis It was a live edit for your screenshot? The best I got is:

![Illustration of the too important spacing between since and types](https://github.com/user-attachments/assets/6fc2574a-21b9-4032-a4a3-80c6840442bd)

Is there any other trick I don't know about Markdown rendering? 😅 (Thanks by the way for the new line explanation! Erika had already told me about it for the MD images inside `figure`... but I'm not used to this syntax so I had forgotten... 👀 )

ArmandPhilippot commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -1308,12 +1479,13 @@ See the `Content Collections` guide for examples of [querying collection entries
 
-<p><Since v="2.5.0" /></p>
+<p>
 
 **Types:**
-
 - `(Array<{ collection: string, id: string }>) => Array<CollectionEntry<collection>>`
 - `(Array<{ collection: string, slug: string }>) => Array<CollectionEntry<collection>>`
+<Since v="2.5.0" />
+</p>
```
Like above: I'm not sure about this one. Since we have a list of types the rendering might seem odd. I tried to place `<Since />` before `Types:` but we have a new line between the two. I think the Markdown parser ignores the `<p>` tag around because of the list inside. So I'm not sure what is the best option here.

delucis replied:
Ah, regarding the list nesting, we might need to use HTML for the whole markup potentially? Which is a bit painful. If indeed `<ul>` is valid inside `<p>`. I think so, but let me double check.

ArmandPhilippot replied:
@delucis https://caninclude.glitch.me/caninclude?child=ul&parent=p It seems we cannot do this. 🫤

delucis replied:
No, `<ul>` is indeed not valid inside a `<p>`, which should only contain phrasing content. Given this, maybe it’s best to flip the order but still not squash everything together into a single paragraph so spacing is consistent.

delucis commented:
Re: https://github.com/withastro/docs/commit/35d1f4b6d342ae740814af0785bf44491a68a202 the newline made sure subsequent markup is treated as Markdown IIRC. In other words the following doesn’t work:
```md
<p>
**Type:** `undefined`
</p>
```
It renders the children as HTML directly:
```html
<p>
**Type:** `undefined`
</p>
```
But the new line lets the Markdown parser take over again:
```md
<p>

**Type:** `undefined`
</p>
```
Renders:
```html
```md
<p>
<strong>Type:</strong> <code>undefined</code>
</p>
```
```

> Even if the method is deprecated, shouldn't it be added?

Yes, probably. I asked the team about the status of `getDataEntryById()` — let’s see what they say!

delucis commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -1490,18 +1696,25 @@ export function onRequest (context, next) {
 
 ### `next()`
```
I wonder if this should be nested under `onRequest()` given it is an argument of that function?

```suggestion
#### `next()`
```

ArmandPhilippot replied:
I checked the history, it seems there is no particular reason for `next()` being at the same level. So I think we can safely move it under `onRequest`. However, it might seem strange to have only one subtitle in the section...

delucis replied:
> it might seem strange to have only one subtitle in the section...

I think it’s OK. There is a question about why `context` doesn’t also have a heading maybe?

ArmandPhilippot replied:
I guess it was tricky since the context object is already described above in `Endpoint contexts`. However, having two subheadings (one for `context` and another for `next()`) since seems preferable to me to avoid an isolated heading (`next()`). Maybe we could use something like:

![An example of context and next function as subheading of onRequest](https://github.com/user-attachments/assets/428aa22f-0897-4a05-9486-5a0ab2c2835c)

**Note:** I didn't check if `Since` could be applied to `context`, this is just an example.

delucis replied:
I like that a lot! And I don’t _think_ we need a `<Since>` for `context` as it was always available in middleware so covered by the main middleware one (and the availability of individual properties should be documented in the endpoint docs you linked to).

(I still don’t think we should display “Added in” in the heading for `next()`, but I guess that was not the point of this screenshot.)

delucis commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -1490,18 +1696,25 @@ export function onRequest (context, next) {
 
-This is an optional argument of `onRequest()`, and may provide the required `Response` returned by the middleware.
+**Type:** `(rewritePayload?: string | URL | Request) => Promise<Response>`<br />
+<Since v="4.13.0" />
```
This one is a bit tricky. There has _always_ been a `next()` since middleware was first added, but the extra `rewritePayload` argument you can pass to it is new since `4.13.0` not quite sure how best to handle this situation. Might make most sense to move the mention inline with the final part about the `rewritePayload` instead of including it here?

ArmandPhilippot replied:
Ah this one is my fault! It's [currently as you describe](https://docs.astro.build/en/reference/api-reference/#next), I haven't checked the history of the argument so I've pulled the information up so it's like the other sections. I'll revert this change!

delucis commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -1490,18 +1696,25 @@ export function onRequest (context, next) {
-**Type:** `(rewritePayload?: string | URL | Request) => Promise<Response>`
+This is an optional argument of `onRequest()`, and may provide the required `Response` returned by the middleware.
 
 It accepts an optional URL path parameter in the form of a string, `URL`, or `Request` to [rewrite](/en/guides/routing/#rewrites) the current request without retriggering a new rendering phase.
```
i.e. something like this:

```suggestion
Since Astro v4.13.0, `next()` accepts an optional URL path parameter in the form of a string, `URL`, or `Request` to [rewrite](/en/guides/routing/#rewrites) the current request without retriggering a new rendering phase.
```

ArmandPhilippot replied:
I didn't read your suggestion before answering to the other comment. This suggestion seems better if we want to keep consistency with the other sections. The [current `Added in:`/`Type:` in the middle of the section](https://docs.astro.build/en/reference/api-reference/#next) seems a little odd.

So should I revert my changes or we use your suggestion?

delucis replied:
I think I like my suggestion 😄 

I agree with you that it’s weird to have the “Added in” in the middle of the paragraph — and even weirder to have the type there. It makes it kind of unclear what is being referred to.

Alternatively, the current style could maybe make more sense with a subheading?

<img width="756" alt="next() API reference section with separate heading for next(requestPayload)" src="https://github.com/user-attachments/assets/3fbc00f3-81fb-4767-aabe-ac6557ba872c">

But I think this overcomplicates things. Keeping the type clearly up top is preferable in my opinion.

ArmandPhilippot replied:
Hum it could be confusing for users I think. They might think that the two are unrelated. I prefer your first suggestion and I think it's better to remove `Since` here. Usually `Since` applies to the section and not the parameter, so something like:

![Example of the restructured next() section](https://github.com/user-attachments/assets/8e665540-0e40-481f-8821-41dd1d565cc3)

ArmandPhilippot replied:
I just realized that this is exactly what you are suggesting when combined with the comment on `Since`. (I must be tired 😅)  So yes, I'm okay with those changes!

delucis reviewed (commented):
Spotted a couple more things reading through the page!

ArmandPhilippot commented:
I think the previous points are all solved!

While I was rereading the page, I noticed that the `getImage` section could be improved. Currently the [`getImage` section in API reference](https://docs.astro.build/en/reference/api-reference/#getimage) is close to a copy-paste of the [section `Generating images with getImage()` in the Images guide](https://docs.astro.build/en/guides/images/#generating-images-with-getimage).

So I added [some changes](https://github.com/withastro/docs/pull/8943/files#diff-42959260ea642887b866e8e276a4520c1f112356eb9c15b31e510132a481aed6R1295-R1340) to differentiate a little the two sections:
* The function signature is now added in API reference (I had hesitated before because of the internal function which has a slightly different signature)
* The returned properties was incomplete, so I completed them
* I thought it might be useful to have the return type here instead of an object, so I updated the code block

I wonder if a link from `Generating images with getImage()` (Images guide) to `getImage()` (API reference) might be helpful. Something like:

```
<ReadMore>Learn more about [the `getImage()` return type](/en/reference/api-reference/#getimage).</ReadMore>
```

But we have almost the same information...

delucis commented:
Nicely spotted! Yes, I’d even say now that once we merge this we could also clean up the “Generating images…” section to be more focused on the practical usage part with a link to reference for the full type details.

delucis commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -1477,7 +1693,9 @@ Middleware allows you to intercept requests and responses and inject behaviors d
-A required exported function from `src/middleware.js` that will be called before rendering every page or API route. It accepts two optional arguments: [context](#contextlocals) and [next()](#next). `onRequest()` must return a `Response`: either directly, or by calling `next()`.
+**Type:** `(context: APIContext, next: MiddlewareNext) => Promise<Response> | Response | Promise<void> | void`
+
+A required exported function from `src/middleware.js` that will be called before rendering every page or API route. It accepts two optional arguments: [context](#context) and [next()](#next). `onRequest()` must return a `Response`: either directly, or by calling `next()`.
```
Small nit regarding the initial wording here:

```suggestion
A required exported function from `src/middleware.js` that will be called before rendering every page or API route. It receives two arguments: [context](#context) and [next()](#next). `onRequest()` must return a `Response`: either directly, or by calling `next()`.
```

Astro always passes the arguments to `onRequest` so they are not “optional arguments”. Whether a user uses them or not is up to them, but they will always be present if they do need them.

delucis commented on src/content/docs/en/reference/api-reference.mdx:
```diff
@@ -1488,19 +1706,36 @@ export function onRequest (context, next) {
+**Type:** `(rewritePayload?: string | URL | Request) => Promise<Response>`<br />
+</p>
 
 A function that intercepts (reads and modifies) the `Response` of a `Request` or calls the "next" middleware in the chain and returns a `Response`. For example, this function could modify the HTML body of a response.
 
 This is an optional argument of `onRequest()`, and may provide the required `Response` returned by the middleware.
```
Not changed in this PR, but taking advantage to suggest some changes as I don’t think this is particularly clear or accurate. This proposal also aligns  the phrasing a bit with the new `context` docs.

```suggestion
The second argument of `onRequest()` is a function that calls all the subsequent middleware in the chain and returns a `Response`. For example, other middleware could modify the HTML body of a response and awaiting the result of `next()` would allow your middleware to respond to those changes.
```

sarah11918 commented:
Looking great, amazing work here! I trust Chris to make the final approval here, and will only say that I agree on scoping *this* PR to the api-reference page only! If you start looking on other pages, I'm sure you'll find more improvements you want to make, so let's make life easier on the translators and tackle these kinds of improvements one page at a time! :smile:

delucis reviewed (commented):
Woo, I think we’re pretty much there — amazing work @ArmandPhilippot! I left a couple of copy suggestions not strictly related to your changes, but thought we might as well get them in while we’re in this file.

sarah11918 commented:
Also pointing out that there's a draft API reference for Actions in #8980 , so feel free to hop into that PR and get that API reference entry up to speed from the beginning!

ArmandPhilippot commented:
Thanks @delucis, the `onRequest` section seems clearer with your two suggestions! (and, while I'm at it, thanks for all your suggestions  ✨   )

I think we've covered this file. There are some missing types but adding them might be more harmful than helpful since they rely on Typescript generics:
* `Astro.glob()` and its overloads
* `props` and `params` in Astro global, context and `getStaticPath`

So unless you see any other improvements, I think we're good this time! 🎉

delucis reviewed (approved):
Agreed! Ready to merge — thanks again for the thorough and persistent work here @ArmandPhilippot 🙌
